### PR TITLE
QAOA evaluate circuit

### PR DESCRIPTION
**Pull request by raulconchello**
## Implementation of `qaoa.evaluate_circuit` method

- [x] New `qaoa.evaluate_circuit` method
  - [x] It accepts list of parameters
  - [x] It accepts dict of parameters
  - [x] If the input different than list or dict, it raises an error
  - [x] The method can be called without optimizing the object only if list or dict is passed
- [x] For variational params, implement `update_from_dict` method, so a dict can be passed to run `evaluate_circuit`
- [x] For variational params, implement `asdict` method, the counterpart of `update_from_dict` method
- [x] tests for all the points above

**Example**

We get some optimized parameters, using -in this case- analytical simulator
```
from openqaoa.problems import QUBO
from openqaoa import QAOA
from openqaoa.backends import create_device
```
```
#create random problem
problem = QUBO.random_instance(6)

#create qaoa object, with analytical simulator
q_analytic = QAOA()
device = create_device(location='local', name='analytical_simulator')
q_analytic.set_device(device)
q_analytic.set_circuit_properties(p=1, param_type='standard', init_type='rand', mixer_hamiltonian='x')
q_analytic.set_classical_optimizer(maxiter=100, method='vgd', jac="finite_difference")

#compile and optimize
q_analytic.compile(problem)
q_analytic.optimize()
```
```` 
opt_params = q_analytic.result.optimized['angles']
print(opt_params)
````

>_**[Out]:** [0.587019236113, 249.611957397334]_

Now we evaluate a qaoa circuit with a QPU or simulator using the optimized parameters found above.
```
#create qaoa object, with shot simulator
q_shots = QAOA()
device = create_device(location='local', name='qiskit.shot_simulator')
q_shots.set_device(device)
q_shots.set_circuit_properties(p=1, param_type='standard', init_type='rand', mixer_hamiltonian='x')
q_shots.set_classical_optimizer(maxiter=100, method='vgd', jac="finite_difference")

#compile (we don't optimize)
q_shots.compile(problem)

#evaluate the qaoa circuit with the optimized parameters
results_eval_circuit = q_shots.evaluate_circuit(opt_params)
print(results_eval_circuit)
```

> _**[Out]:** {
>            'cost': -44.783801067124,
>  'uncertainty': 103.578710845481,
>  'counts': {...}
> }

### CODE REVIEW

File: tests/test_workflows.py [hunk 1]
```diff
@@ -1100,6 +1103,137 @@ def test_qaoa_from_dict_and_load(self):
...
+            q.compile(maxcut_qubo) 
+            qaoas.append(q)
+
+        # for each qaoa object, test the evaluate_circuit method
+        for q in qaoas:
+
+            # evaluate the circuit with random dict of params
+            params = {k: np.random.rand(*v.shape) for k, v in q.variate_params.asdict().items()}
+            result = q.evaluate_circuit(params)
+            assert abs(result['cost']) > 0, f"param_type={q.circuit_properties.param_type}. `evaluate_circuit` should return a cost, here cost is {result['cost']}"
```
Comment (Q-lds) [inline tests/test_workflows.py:1128]: The cost can also be negative, no? Maybe you can let the maxcut problem to be a random problem, so we catch edge cases?

Comment (raulconchello) [reply tests/test_workflows.py:1128]: yes I put `abs(cost)`, it could be also 0 but very unlikely

Comment (raulconchello) [reply tests/test_workflows.py:1128]: I changed it to >= and random qubo

File: src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py [hunk 2]
```diff
@@ -303,6 +307,74 @@ def optimize(self, verbose=False):
...
+
+        # before evaluating the circuit we check that the QAOA object has been compiled
+        if self.compiled == False: raise ValueError("Please compile the QAOA before optimizing it!")
+
+        
+        ## Check the type of the input parameters and save them as a QAOAVariationalBaseParams object at the variable `params_obj`
+
+        # if no parameters are passed, we use the optimized parameters of the QAOA object
+        if params is None:
+            assert self.result != None, "No parameters have been passed and the QAOA object has not been optimized yet"
```
Comment (vishal-ph) [inline src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py:341]: We are not performing optimization here, so shouldn't we change the error message?

File: src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py [hunk 3]
```diff
@@ -303,6 +307,74 @@ def optimize(self, verbose=False):
...
+
+        # if the parameters are passed as a list we copy and update the variational parameters of the QAOA object
+        elif isinstance(params, list) or isinstance(params, np.ndarray):
+            assert len(params) == len(self.variate_params), "The number of parameters does not match the number of parameters in the QAOA circuit"
+            params_obj = deepcopy(self.variate_params)
+            params_obj.update_from_raw(params)
+
+        # if the parameters are passed in a different format, we raise an error
+        else:
+            raise TypeError("`params` must be a list, a dictionary or None")
```
[the remark this answers is not in the document]
Comment (Q-lds) [reply src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py:362]: I think that for a `TypeError` it also makes sense to return what type you passed 

`raise TypeError(f"`params` must be a list, a dictionary or None". Instead, received {var})`

[the remark this answers is not in the document]
Comment (raulconchello) [reply src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py]: done

File: src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py (hunk 2 above)
Comment (raulconchello) [reply src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py]: changed, now the input must be list or dict

File: src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py [hunk 4]
```diff
@@ -303,6 +308,82 @@ def optimize(self, verbose=False):
         if verbose:
             print(f"optimization completed.")
         return
+    
+    def evaluate_circuit(
+        self, params: Union[List[float], Dict[str, List[float]]]
```
Comment (raulconchello) [inline src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py:313]: Question: Should you be able to pass a QAOAVariationalBaseParams object too?

Comment (raulconchello) [reply src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py]: I have updated that, but the problem that I see is that you could pass an object with different properties than the QAOA object compiled. For example:
```
q = QAOA()
q.compile(problem)

params_obj = (create a params obj with p>1)
q.evaluate_circuit(params_obj)
```

File: src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py [hunk 5]
```diff
@@ -303,6 +309,92 @@ def optimize(self, verbose=False):
...
+        if isinstance(self.backend, QAOABaseBackendStatevector):
+            state = self.backend.wavefunction(params_obj)
+            cost, uncertainty = self.backend.expectation_w_uncertainty(params_obj)
+            return {
+                'cost': cost,
+                'uncertainty': uncertainty,
+                'state': state
+            }
+
+        # if the backend is a QPU or a shot-based simulator, we return the expectation value and the uncertainty of the cost Hamiltonian and the counts
```
Comment (vishal-ph) [inline src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py:381]: why don't we use `expectation_w_uncertainty` method for non-statevector backends too?

**Comment by Q-lds** [comment]
Let's remember to close Issue https://github.com/entropicalabs/openqaoa/issues/186 once this PR is merged!

File: src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py (hunk 4 above)
Comment (vishal-ph) [reply src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py]: I guess we can just raise an error for this case.

**Comment by vishal-ph** [comment]
NOTE: run `black` formatter before merging

File: src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py (hunk 4 above)
Comment (vishal-ph) [reply src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py]: I have added some checks for this

**Review by vishal-ph — APPROVED** [review]
